cpp/oneapi/dal/table/detail/access_iface.hpp

```diff
@@ -21,81 +21,275 @@
+#define DECLARE_PUSH_ROWS_HOST(T)   virtual PUSH_ROWS_SIGNATURE_HOST(T) = 0;
+#define DECLARE_PUSH_ROWS_DPC(T)    virtual PUSH_ROWS_SIGNATURE_DPC(T) = 0;
+#define DECLARE_PUSH_COLUMN_HOST(T) virtual PUSH_COLUMN_SIGNATURE_HOST(T) = 0;
+#define DECLARE_PUSH_COLUMN_DPC(T)  virtual PUSH_COLUMN_SIGNATURE_DPC(T) = 0;
+
+#define DEFINE_TEMPLATE_PULL_ROWS_HOST(T)                                 \
+    PULL_ROWS_SIGNATURE_HOST(T) override {                                \
+        static_cast<Derived*>(this)->pull_rows(policy, block, row_range); \
+    }
+
+#define DEFINE_TEMPLATE_PULL_ROWS_DPC(T)                                         \
+    PULL_ROWS_SIGNATURE_DPC(T) override {                                        \
+        static_cast<Derived*>(this)->pull_rows(policy, block, row_range, alloc); \
+    }
+
+#define DEFINE_TEMPLATE_PULL_COLUMN_HOST(T)                                               \
+    PULL_COLUMN_SIGNATURE_HOST(T) override {                                              \
+        static_cast<Derived*>(this)->pull_column(policy, block, column_index, row_range); \
+    }
+
+#define DEFINE_TEMPLATE_PULL_COLUMN_DPC(T)                                                       \
+    PULL_COLUMN_SIGNATURE_DPC(T) override {                                                      \
+        static_cast<Derived*>(this)->pull_column(policy, block, column_index, row_range, alloc); \
+    }
+
+#define DEFINE_TEMPLATE_PUSH_ROWS_HOST(T)                                 \
+    PUSH_ROWS_SIGNATURE_HOST(T) override {                                \
+        static_cast<Derived*>(this)->push_rows(policy, block, row_range); \
+    }
+
+#define DEFINE_TEMPLATE_PUSH_ROWS_DPC(T)                                  \
+    PUSH_ROWS_SIGNATURE_DPC(T) override {                                 \
+        static_cast<Derived*>(this)->push_rows(policy, block, row_range); \
+    }
+
+#define DEFINE_TEMPLATE_PUSH_COLUMN_HOST(T)                                               \
+    PUSH_COLUMN_SIGNATURE_HOST(T) override {                                              \
+        static_cast<Derived*>(this)->push_column(policy, block, column_index, row_range); \
+    }
+
+#define DEFINE_TEMPLATE_PUSH_COLUMN_DPC(T)                                                \
+    PUSH_COLUMN_SIGNATURE_DPC(T) override {                                               \
+        static_cast<Derived*>(this)->push_column(policy, block, column_index, row_range); \
+    }
+
+class pull_rows_iface {
+public:
+    virtual ~pull_rows_iface() = default;
+
+    DECLARE_PULL_ROWS_HOST(float)
+    DECLARE_PULL_ROWS_HOST(double)
+    DECLARE_PULL_ROWS_HOST(std::int32_t)
+
+#ifdef ONEDAL_DATA_PARALLEL
+    DECLARE_PULL_ROWS_DPC(float)
+    DECLARE_PULL_ROWS_DPC(double)
+    DECLARE_PULL_ROWS_DPC(std::int32_t)
+#endif
 };
 
-struct column_values_block {
-    std::int64_t column_index;
-    range rows;
+template <typename Derived>
+class pull_rows_template : public base, public pull_rows_iface {
```

Comment:
1. It is not clear how it depends on `Derived`. Probably it is better to pass it as a parameter to defines `DEFINE_TEMPLATE_PULL_ROWS_*`?
2. Why do we really need to have such `*_template` helpers for pull/push and tables and cannot just inherit our implementations from interfaces directly? Is it recommended to create such a template for every new access iface or table?

Reply:
> 1. It is not clear how it depends on Derived. Probably it is better to pass it as a parameter to defines DEFINE_TEMPLATE_PULL_ROWS_*?

ok

----------------------------------------
> 2. Why do we really need to have such *_template helpers for pull/push and tables and cannot just inherit our implementations from interfaces directly? Is it recommended to create such a template for every new access iface or table?

The reason is multiple pull/push overloads that accepts float/double/int. If I inherit from interface directly, I'll need to explicitly define all those overloads. `*_template` helpers allows using template functions by means of CRTP magic.

**CRTP**
```c++
class my_table : public pull_rows_template<my_table> {
public:
   template <typename Policy, typename T>
   void pull(const Policy& policy, array<T>& block) { ... }
};
``` 

**"Direct"**
```c++
class my_table : public pull_rows_iface {
public:
   void pull(const default_host_policy& policy, array<float>& block) override { ... }
   void pull(const default_host_policy& policy, array<double>& block) override { ... }
   void pull(const default_host_policy& policy, array<std::int32_t>& block) override { ... }
   void pull(const data_parallel_policy& policy, array<float>& block) override { ... }
   void pull(const data_parallel_policy& policy, array<double>& block) override { ... }
   void pull(const data_parallel_policy& policy, array<std::int32_t>& block) override { ... }

protected:
   template <typename Policy, typename T>
   void pull_impl(...) { ... }
};
``` 

I'll need to repeat this in each table impl

Reply:
Thank you